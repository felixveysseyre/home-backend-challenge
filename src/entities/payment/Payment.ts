export type Payment = {
  id: string;
  contractId: string;
  description: string;
  value: number;
  time: Date;
  createdAt: Date;
  updatedAt: Date | undefined;
  deletedAt: Date | undefined;
};
