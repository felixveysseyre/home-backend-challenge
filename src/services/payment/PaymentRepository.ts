import { Injectable } from '@nestjs/common';
import { nanoid } from 'nanoid';

import { Payment } from '../../entities/Payment';

@Injectable()
export class PaymentRepository {
  private store: Payment[] = [];

  public add({
    payment,
  }: {
    payment: Omit<Payment, 'id' | 'createdAt' | 'updatedAt' | 'deletedAt'>;
  }) {
    const paymentCreated = {
      ...payment,
      id: nanoid(),
      createdAt: new Date(),
      updatedAt: undefined,
      deletedAt: undefined,
    };

    this.store.push(paymentCreated);

    return paymentCreated;
  }

  public get({
    ids,
    contractIds,
    timeInterval,
  }: {
    ids?: Payment['id'][];
    contractIds?: Payment['contractId'][];
    timeInterval?: {
      after?: Payment['time'];
      before?: Payment['time'];
    };
  }) {
    /* Checkers */

    const checkId = (payment: Payment) => {
      return ids ? ids.includes(payment.id) : true;
    };

    const checkContractId = (payment: Payment) => {
      return contractIds ? contractIds.includes(payment.contractId) : true;
    };

    const checkTime = (payment: Payment) => {
      if (!timeInterval) {
        return true;
      }

      if (timeInterval.after && timeInterval.before) {
        return (
          payment.time.valueOf() >= timeInterval.after.valueOf() &&
          payment.time.valueOf() <= timeInterval.before.valueOf()
        );
      }

      if (timeInterval.after) {
        return payment.time.valueOf() >= timeInterval.after.valueOf();
      }

      if (timeInterval.before) {
        return payment.time.valueOf() <= timeInterval.before.valueOf();
      }
    };

    /* Return */

    return this.store.filter((payment) => {
      return checkId(payment) && checkContractId(payment) && checkTime(payment);
    });
  }

  public getById({ id }: { id: Payment['id'] }) {
    const payments = this.get({ ids: [id] });

    if (payments.length !== 1) {
      throw new Error(`Payment not found: ${id}`);
    }

    return payments[0];
  }

  public update({
    update,
  }: {
    update: Pick<Payment, 'id'> & Partial<Omit<Payment, 'id'>>;
  }) {
    const paymentFound = this.getById({ id: update.id });
    const paymentUpdated = {
      ...paymentFound,
      updatedAt: new Date(),
      ...update,
    };

    this.store = this.store.map((payment) => {
      return payment.id === update.id ? paymentUpdated : payment;
    });

    return paymentUpdated;
  }

  public delete({ id }: { id: Payment['id'] }) {
    return this.update({ update: { id, deletedAt: new Date() } });
  }

  private clear() {
    this.store = [];
  }
}
