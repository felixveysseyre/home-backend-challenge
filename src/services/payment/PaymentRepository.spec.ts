import { Test } from '@nestjs/testing';
import { appModuleMetadata } from '../../app.module';
import { Payment } from '../../entities/Payment';

import { PaymentRepository } from './PaymentRepository';

describe('PaymentRepository', () => {
  let paymentRepository: PaymentRepository;

  beforeAll(async () => {
    const module = await Test.createTestingModule(appModuleMetadata).compile();

    paymentRepository = module.get<PaymentRepository>(PaymentRepository);
  });

  beforeEach(() => {
    // @ts-expect-error: Method is private
    paymentRepository.clear();
  });

  describe('add', () => {
    it('succeeds to add a payment', () => {
      const paymentCreated = paymentRepository.add({
        payment: {
          contractId: 'fakeContractId',
          description: 'Fake description',
          value: 123,
          time: new Date(),
        },
      });

      expect(paymentCreated).toMatchSnapshot({
        id: expect.any(String),
        time: expect.any(Date),
        createdAt: expect.any(Date),
      });
    });
  });

  describe('get', () => {
    let payments: Payment[];

    beforeEach(() => {
      payments = [
        paymentRepository.add({
          payment: {
            contractId: 'fakeContractId1',
            description: 'Fake description',
            value: 123,
            time: new Date(1),
          },
        }),
        paymentRepository.add({
          payment: {
            contractId: 'fakeContractId2',
            description: 'Fake description',
            value: 123,
            time: new Date(2),
          },
        }),
      ];
    });

    it('succeeds to get payments, filtered by id', () => {
      expect(paymentRepository.get({ ids: [payments[0].id] })).toEqual([
        payments[0],
      ]);
    });

    it('succeeds to get payments, filtered by contractId', () => {
      expect(
        paymentRepository.get({ contractIds: [payments[0].contractId] }),
      ).toEqual([payments[0]]);
    });

    it('succeeds to get payments, filtered by timeInterval after', () => {
      expect(
        paymentRepository.get({ timeInterval: { after: payments[1].time } }),
      ).toEqual([payments[1]]);
    });

    it('succeeds to get payments, filtered by timeInterval before', () => {
      expect(
        paymentRepository.get({ timeInterval: { before: payments[0].time } }),
      ).toEqual([payments[0]]);
    });

    it('succeeds to get payments, filtered by timeInterval after and before', () => {
      expect(
        paymentRepository.get({
          timeInterval: { after: payments[0].time, before: payments[1].time },
        }),
      ).toEqual(payments);
    });
  });

  describe('getById', () => {
    let payment: Payment;

    beforeEach(() => {
      payment = paymentRepository.add({
        payment: {
          contractId: 'fakeContractId',
          description: 'Fake description',
          value: 123,
          time: new Date(1),
        },
      });
    });

    it('succeeds to get payment', () => {
      expect(paymentRepository.getById({ id: payment.id })).toEqual(payment);
    });

    it('fails to get payment, if not existing', () => {
      expect(() => paymentRepository.getById({ id: 'falseId' })).toThrow(
        'Payment not found',
      );
    });
  });

  describe('update', () => {
    let payment: Payment;

    beforeEach(() => {
      payment = paymentRepository.add({
        payment: {
          contractId: 'fakeContractId',
          description: 'Fake description',
          value: 123,
          time: new Date(1),
        },
      });
    });

    it('succeeds to update payment', () => {
      const update = {
        id: payment.id,
        contractId: 'updateFakeContractId',
        description: 'Updated fake description',
        value: 1234,
        time: new Date(2),
      };

      const updatedPayment = paymentRepository.update({ update });

      expect(updatedPayment.id).toBe(payment.id);
      expect(updatedPayment).toMatchSnapshot({
        id: expect.any(String),
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
      });

      expect(paymentRepository.getById({ id: payment.id })).toEqual(
        updatedPayment,
      );
    });

    it('fails to update payment, if not existing', () => {
      expect(() =>
        paymentRepository.update({
          update: {
            id: 'falseId',
          },
        }),
      ).toThrow('Payment not found');
    });
  });

  describe('delete', () => {
    let payment: Payment;

    beforeEach(() => {
      payment = paymentRepository.add({
        payment: {
          contractId: 'fakeContractId',
          description: 'Fake description',
          value: 123,
          time: new Date(1),
        },
      });
    });

    it('succeed to delete a payment', () => {
      const deletedPayment = paymentRepository.delete({ id: payment.id });

      expect(deletedPayment.id).toBe(payment.id);
      expect(deletedPayment).toMatchSnapshot({
        id: expect.any(String),
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
        deletedAt: expect.any(Date),
      });

      expect(paymentRepository.getById({ id: payment.id })).toEqual(
        deletedPayment,
      );
    });

    it('fails to delete payment, if not existing', () => {
      expect(() =>
        paymentRepository.delete({
          id: 'falseId',
        }),
      ).toThrow('Payment not found');
    });
  });
});
