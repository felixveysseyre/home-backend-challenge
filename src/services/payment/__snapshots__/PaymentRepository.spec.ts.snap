// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaymentRepository add succeeds to add a payment 1`] = `
Object {
  "contractId": "fakeContractId",
  "createdAt": Any<Date>,
  "deletedAt": undefined,
  "description": "Fake description",
  "id": Any<String>,
  "time": Any<Date>,
  "updatedAt": undefined,
  "value": 123,
}
`;

exports[`PaymentRepository delete succeed to delete a payment 1`] = `
Object {
  "contractId": "fakeContractId",
  "createdAt": Any<Date>,
  "deletedAt": Any<Date>,
  "description": "Fake description",
  "id": Any<String>,
  "time": 1970-01-01T00:00:00.001Z,
  "updatedAt": Any<Date>,
  "value": 123,
}
`;

exports[`PaymentRepository update succeeds to update payment 1`] = `
Object {
  "contractId": "updateFakeContractId",
  "createdAt": Any<Date>,
  "deletedAt": undefined,
  "description": "Updated fake description",
  "id": Any<String>,
  "time": 1970-01-01T00:00:00.002Z,
  "updatedAt": Any<Date>,
  "value": 1234,
}
`;
